lessons/misc/intro-bash/lesson.md

```diff
@@ -0,0 +1,354 @@
+
+```
+grep '^E' animals.txt >> fish.txt && mv fish.txt fish-plus-E.txt
+```
+
+Note the use of `&&`. This allows us to run multiple commands in a single line. In fact, there are a few such commands, which I'll list here:
+
+- `;` separates commands like they are on separate lines
+- `&&` does the next command if the previous one succeeds
+- `||` does the next command if the previous one fails
+- `&` runs the previous command in the "background"
+
+# Running programs from the command-line
+
+It can be quite useful to write entire scripts and run them in their entirety. This is particularly helpful if running scripts (e.g. in R, Python, etc) on a server or coputer cluster where Graphical User Interfaces (GUI's) are uncommon.
+
+Before starting, let's change into our `scripts` folder. Note, we could run the scripts from aany folder by specifying the full path to the file, but for now we'll change over.
+
+```
+cd ../scripts
+```
+
+We'll be executing the python script `hello_world.py`. First, let's make sure that we have Python and that the computer knows where to find it (more on this in a bit). `which` is a helpful command for this as it tells us which *executable file* is being run when we type a command in the shell.
+
+```
+which python
+```
+
+If successful, the shell will print the full path to the executable used in performing the command. Incidentally, these are the files that are executed when your using your computer for everyday use; they're execution is simply hidden because you're often using a GUI.
+
+Let's go ahead and run our script
+
+```
+python hello_world.py
+```
+
+We can similarly execute bash scripts in the same way.
+
+```
+bash print-numbers.sh
+```
+
+R scripts can be run using the `Rscript <script-name>` command.
+
+# Environment variables
+
+There are special variables called *environment variables* which are used by bash and other programs to affect behaviour. Here, we'll only cover two of them. The first is one of the most important: `$PATH`:
+
+```
+echo $PATH
+```
+
+That's a little tough to read. Let's put the directories in our `$PATH` on separate lines by piping the output of `$PATH` to the **tr**anslate characters command, which allows us to quickly substitude the `:` in `$PATH` with newline (`\n`) characters.
+
+```
+echo "$PATH" | tr ':' '\n'
+```
+
+Notice the `$` symbol. This is how variables are accessed in bash. *Environment variables* are always capitalized. The `$PATH` variable sotres the directories you computer searches for programs, commands, and utilitis that have been installed. If you installed something and can't get it to work on the command-line, it's likely because the *executable file* for the program is not in the computers `$PATH`. When we used the `which` command earlier, these are the directories that were searched. If the command in not in `$PATH`, nothing will be printed.
```

Comment:
s/sotres/stores

s/utilitis/utilities

s/the command in not/the command is not